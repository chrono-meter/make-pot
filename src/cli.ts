#!/usr/bin/env node

import process from "node:process";
import * as yargs from "yargs";
import { hideBin } from "yargs/helpers";
import { getArgs } from "./cli/getArgs.js";
import { getJsonArgs } from "./cli/getJsonArgs.js";
import makeJsonCommand from "./jsonCommand.js";
import makepotCommand from "./potCommand.js";
import type { Args, MakeJsonArgs } from "./types.js";

/** Main execution */
// Get the selected command
const r = yargs
	.default(hideBin(process.argv))
	.options({
		makejson: {
			describe: "Make JSON file",
			type: "boolean",
			default: false,
		},
	})
	.parseSync() as { makejson: boolean };

// Execute the command
if (!r.makejson) {
	makepotCommand(getArgs() as Args);
} else {
	makeJsonCommand(getJsonArgs() as MakeJsonArgs);
}
