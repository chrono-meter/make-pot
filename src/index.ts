#!/usr/bin/env node
import { makePot } from './makePot'

import { getArgs } from './cliArgs'

import yargs from 'yargs'

import { name, version } from '../package.json'

/** Main execution */
const args = getArgs()
if (Object.keys(args).length > 0) {
	console.log(name + ' version: ' + version)
	const timeStart = new Date()
	makePot(args)
		.then(() => {
			const timeEnd = new Date()
			console.log(
				`ðŸš€ Translation Pot file created in ${timeEnd.getTime() - timeStart.getTime()}ms`
			)
		})
		.catch((error) => {
			console.error(error)
		})
} else {
	yargs.showHelp()
}
