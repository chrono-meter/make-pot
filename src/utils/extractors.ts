import { Block, SetOfBlocks } from "gettext-merger";

/**
 * Returns the key of an object based on its value
 *
 * @param object the object that contains the key
 * @param value the key that we want to get
 * @return {Record<string, string>} the filtered keys
 */
export function getKeyByValue(
	object: Record<string, unknown>,
	value: string,
): string | undefined {
	return Object.keys(object).find((key) => object[key] === value) ?? undefined;
}

/**
 * Returns a gettext translation object
 *
 * @param label the label of the translation
 * @param string the string of the translation
 * @param filePath the file path of the translation
 */
export const buildBlock = (
	label: string,
	string: string,
	filePath: string[] | undefined = undefined,
): Block => {
	const block = new Block([]);
	block.msgctxt = undefined;
	block.msgid = string;
	block.msgid_plural = "";
	block.msgstr = [];
	block.comments = {};
	if (label) {
		block.comments.extracted = [label];
	}
	if (filePath?.length) {
		block.comments.reference = filePath;
	}
	return block;
};

/**
 * Extracts strings from parsed JSON data.
 *
 * @param {Record<string, any> | Parser.SyntaxNode} parsed - The parsed JSON data or syntax node.
 * @param {string | Parser} filename - The filename or parser.
 * @param filepath - the path to the file being parsed
 * @return {SetOfBlocks} An array of translation strings.
 */
export function yieldParsedData(
	parsed: Block[],
	filename: "block.json" | "theme.json",
	filepath: string,
): SetOfBlocks {
	const gettextTranslations: SetOfBlocks = new SetOfBlocks([], filepath);

	if (parsed.length === 0) {
		return gettextTranslations;
	}

	// set the path of the translation
	gettextTranslations.path = filepath;

	for (const item of parsed) {
		const block = buildBlock(
			item.msgid,
			item.msgctxt as string,
			item.comments?.reference,
		);

		if (block) {
			gettextTranslations.blocks.push(block);
		}
	}

	return gettextTranslations;
}
