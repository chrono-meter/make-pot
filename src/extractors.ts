import path from 'path'
import fs, { readFileSync } from 'fs'
import { type Args, type TranslationString } from './types'
import { pkgJsonHeaders } from './const'
import { getCommentBlock } from './utils'
import Parser from 'tree-sitter'
import type { SingleBar } from 'cli-progress'
import { jsonString, parseJsonFile } from './extractors-json'
import { parsePHPFile } from './extractors-php'
import { extractStrings } from './tree'
import { extractFileData } from './extractors-text'

/**
 * Extracts the names from an array of items.
 *
 * @param {unknown[]} items - The array of items.
 * @return {string[]} An array of names extracted from the items.
 */
export const extractNames = (items: { name: string }[]): string[] => items.map((item) => item.name)

/**
 * Extracts strings from parsed JSON data.
 *
 * @param {Record<string, any> | Parser.SyntaxNode} parsed - The parsed JSON data or syntax node.
 * @param {string | Parser} filename - The filename or parser.
 * @param opts - The metadata of this translation string.
 * @return {TranslationString[]} An array of translation strings.
 */
export function yieldParsedData(
	parsed: Record<string, string | string[]> | Parser.SyntaxNode,
	filename: string | Parser,
	opts: { filepath: string; stats?: { bar: SingleBar } }
): Promise<TranslationString[]> {
	return new Promise<TranslationString[]>((resolve) =>
		resolve(
			Object.entries(parsed as Record<string, string | string[]>)

				// return the translations for each key in the json data
				.map(([key, jsonData]) => {
					// if is a string return a single json string
					if (typeof jsonData === 'string') {
						return jsonString(
							key,
							jsonData,
							opts.filepath,
							filename as 'block.json' | 'theme.json'
						)
					} else {
						opts.stats?.bar.increment(0, { filename: 'not a string' })
					}

					if (!jsonData) {
						opts.stats?.bar.increment(0, {
							filename: `Skipping ${key} in ${opts.filepath} as ${filename} ... cannot parse data`,
						})

						return null
					}

					// if is an object return an array of json strings
					Object.entries(jsonData).map(([k, v]) => jsonString(k, v, opts.filepath))
				})
				.flat() as TranslationString[]
		)
	)
}

/**
 * Parses the source code using the specified language parser and extracts the strings from the file.
 *
 * @param {string} sourceCode - The source code to be parsed.
 * @param {Args} language - The language to be used for parsing.
 * @param {string} filepath - The path to the file being parsed.
 * @return {TranslationString[]} An array of translation strings.
 */
export function doTree(sourceCode: string, language: Parser, filepath: string) {
	// set up the parser
	const parser = new Parser()
	parser.setLanguage(language)

	// parse the file
	const tree = parser.parse(sourceCode) // Assuming parse is an async operation

	// extract the strings from the file and return them
	return extractStrings(tree.rootNode, language, filepath)
}

/**
 * Parse a file and extract strings asynchronously
 *
 * @param {object} args
 * @param {string} args.filepath - Path to the file to parse
 * @param {Parser|null} args.language - Language of the file to parse
 * @return {Promise<TranslationString[]>}
 */
export async function parseFile(args: {
	filepath: string
	language: Parser | string
}): Promise<TranslationString[] | null> {
	// check if the language is supported
	if (typeof args.language === 'string') {
		if (args.language === 'json') {
			return parseJsonFile(args) || []
		}
		console.log(`Skipping ${args.filepath}... No parser found for ${args.language} file`)
		return null
	}

	// read the file
	const sourceCode = readFileSync(path.resolve(args.filepath), 'utf8')

	// set up the parser
	const parser = new Parser()
	parser.setLanguage(args.language)

	// parse the file
	const tree = parser.parse(sourceCode) // Assuming parse is an async operation

	// extract the strings from the file and return them
	return extractStrings(tree.rootNode, args.language, args.filepath)
}

/**
 * Extracts package data from the given arguments and returns a record
 * containing the specified fields from the package.json file.
 *
 * @param {Args} args - The arguments for extracting package data.
 * @param {Record<string, string>} fields - The fields to extract from the package.json file. Default is pkgJsonHeaders.
 * @return {Record<string, string>} - A record containing the extracted package data.
 */
export function extractPackageData(args: Args, fields = pkgJsonHeaders): Record<string, string> {
	// TODO: package.json "files" could be used to get the file list
	const pkgJsonMeta: Record<string, string> = {}
	// read the package.json file
	const packageJsonPath = args.sourceDirectory
		? path.join(args.sourceDirectory, 'package.json')
		: 'package.json'
	if (fs.existsSync(packageJsonPath)) {
		const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'))
		// extract the fields from the package.json file
		for (const field of Object.keys(fields)) {
			if (packageJson[field] !== undefined) {
				pkgJsonMeta[field] = packageJson[field]
			}
		}
	}
	return pkgJsonMeta
}

/**
 * Extracts main file data based on the given arguments.
 *
 * @param {Args} args - The arguments for extracting the main file data.
 * @return {Record<string, string>} The extracted main file data.
 */
export function extractMainFileData(args: Args) {
	let fileData: Record<string, string> = {}
	const sourceDir = args.sourceDirectory
		? path.join(process.cwd(), args.sourceDirectory)
		: process.cwd()

	if (['plugin', 'block', 'generic'].includes(args.domain)) {
		const folderPhpFile = path.join(sourceDir, `${args.slug}.php`)

		if (fs.existsSync(folderPhpFile)) {
			const fileContent = fs.readFileSync(folderPhpFile, 'utf8')
			fileData = parsePHPFile(fileContent)

			if ('Plugin Name' in fileData) {
				console.log('Plugin file detected.')
				console.log(`Plugin file: ${folderPhpFile}`)
				args.domain = 'plugin'
			}
		} else {
			console.log('Plugin file not found.')
			console.log(`Missing Plugin filename: ${folderPhpFile}`)
		}
	} else if (['theme', 'theme-block'].includes(args.domain)) {
		const styleCssFile = path.join(sourceDir, 'style.css')

		if (fs.existsSync(styleCssFile)) {
			const fileContent = fs.readFileSync(styleCssFile, 'utf8')
			const commentBlock = getCommentBlock(fileContent)
			fileData = extractFileData(commentBlock)

			console.log('Theme stylesheet detected.')
			console.log(`Theme stylesheet: ${styleCssFile}`)
			args.domain = 'theme'
		} else {
			console.log(`Theme stylesheet not found in ${path.resolve(sourceDir)}`)
		}
	}

	return fileData
}
