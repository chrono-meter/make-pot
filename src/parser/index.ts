import { type Args } from '../types'
import { extractMainFileData } from '../extractors/headers'
import { writeFile } from '../fs'
import path from 'path'
import { extractPackageJson } from '../extractors/json'
import { exec } from './exec'

function getOutputPath(args: Args): string {
	return path.join(
		process.cwd(),
		(args.headers?.domainPath as string) ?? args.paths.out ?? 'languages',
		`${args?.slug}.${args.options?.json ? 'json' : 'pot'}`
	)
}

/**
 * Generates a pot file for localization.
 *
 * @param args - the command line arguments
 * @return {Promise<void>} - a promise that resolves when the pot file is generated
 */
export async function makePot(args: Args): Promise<string> {
	/** Collect metadata from the get package json */
	const pkgData = extractPackageJson(args)

	/** Get metadata from the main file (theme and plugin) */
	const metadata = extractMainFileData(args)

	/** Merge the metadata to get a single object with all the headers */
	args.headers = {
		...args.headers,
		...pkgData,
		...metadata,
	} as Args['headers']

	/** Generate the pot file */
	const jsonTranslations = await exec(args)

	const outputPath = getOutputPath(args)

	console.log(`Writing pot file to ${outputPath}`)

	writeFile(jsonTranslations, outputPath)

	return jsonTranslations
}
