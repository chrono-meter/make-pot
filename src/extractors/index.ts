import * as path from 'path'
import { readFileAsync } from '../fs'
import { parseJsonCallback } from './json'
import { doTree } from '../parser/tree'
import { SetOfBlocks } from 'gettext-merger'
import { allowedFiles } from '../const'

/**
 * Parse a file and extract strings asynchronously
 *
 * @return {Promise<SetOfBlocks>}
 */
export async function parseFile(
	file: string,
	filePath: string = ''
): Promise<SetOfBlocks | undefined> {
	const filename = path.basename(file)
	const ext = path.extname(file).replace(/^./, '')
	const fileRealPath = path.resolve(filePath, file)

	if (ext === 'json') {
		if (filename === 'theme.json' || filename === 'block.json') {
			return readFileAsync(fileRealPath).then(async (sourceCode) => {
				return await parseJsonCallback(sourceCode, filePath, filename)
			})
		}
	}

	if (allowedFiles.includes(ext)) {
		// read the file
		return readFileAsync(fileRealPath).then((sourceCode) => {
			return doTree(sourceCode, file)
		})
	}
}
