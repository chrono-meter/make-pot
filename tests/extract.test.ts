import { describe, expect } from '@jest/globals'
import { getStrings } from '../src/parser'
import { doTree } from '../src/extractors'
// @ts-expect-error
import Js from 'tree-sitter-javascript'
// @ts-expect-error
import Php from 'tree-sitter-php'
// @ts-expect-error
import Ts from 'tree-sitter-typescript'
import Parser from 'tree-sitter'

describe('getStrings', () => {
	it('should extract translations with context', () => {
		const filename = 'filename'
		const content = `<?php echo _x('Hello World', 'greeting'); ?>`
		const expected = [
			{
				comments: undefined,
				reference: '#: filename:1',
				type: 'text_context_domain',
				raw: ['Hello World', 'greeting'],
				msgid: 'Hello World',
			},
		]

		const result = doTree(content, Php as Parser, filename)

		expect(result).toEqual(expected)
	})
	it('should extract translations from code content with no context or translator comments', () => {
		const filename = 'filename'
		const content = `<?php _e('Hello World'); ?>`
		const expected = [
			{
				reference: '#: filename:1',
				type: 'text_domain',
				raw: ['Hello World'],
				msgid: 'Hello World',
			},
		]

		const result = doTree(content, Php as Parser, filename)

		expect(result).toEqual(expected)
	})

	it('should extract translations with comments', () => {
		const filename = 'filename'
		const content = `/** translators: ciao! */
		<?php echo _x('Hello World', 'greeting'); ?>`
		const expected = [
			{
				comments: 'translators: ciao!',
				reference: '#: filename:2',
				type: 'text_context_domain',
				raw: ['Hello World', 'greeting'],
				msgid: 'Hello World',
			},
		]

		const result = doTree(content, Php as Parser, filename)

		expect(result).toEqual(expected)
	})

	it('should extract translations with comments reporting the right position', () => {
		const filename = 'filename'
		const content = `



/** line 5*/




		<?php echo _x('Hello World', 'greeting'); ?>`
		const expected = [
			{
				reference: '#: filename:10',
				type: 'text_context_domain',
				raw: ['Hello World', 'greeting'],
				msgid: 'Hello World',
			},
		]

		const result = doTree(content, Php as Parser, filename)

		expect(result).toEqual(expected)
	})

	it('should extract translations inside a sprint', () => {
		const filename = 'filename'
		const content = ` <?php
$url = 'http://example.com';
$link = sprintf( wp_kses( __( 'Check out this link to my <a href="%s">website</a> made with WordPress.', 'my-text-domain' ), array(  'a' => array( 'href' => array() ) ) ), esc_url( $url ) );
echo $link;`
		const expected = [
			{
				msgid: 'Check out this link to my <a href="%s">website</a> made with WordPress.',
				raw: [
					'Check out this link to my <a href="%s">website</a> made with WordPress.',
					'my-text-domain',
				],
				reference: '#: filename:3',
				type: 'text_domain',
			},
		]

		const result = doTree(content, Php as Parser, filename)

		expect(result).toEqual(expected)
	})

	it('should extract translations with translator comments inside the formatting hell', () => {
		const filename = 'filename'
		const content = ` <?php /** 1*/
/** 2*/
/** 3*/		printf(
/** 4*/			/* translators: 1: Site URL, 2: Username, 3: User email address, 4: Lost password URL. */
		__( 'Your site at %1$s is active. You may now log in to your site using your chosen username of &#8220;%2$s&#8221;. Please check your email inbox at %3$s for your password and login instructions. If you do not receive an email, please check your junk or spam folder. If you still do not receive an email within an hour, you can <a href="%4$s">reset your password</a>.' ),
/** 6*/			sprintf( '<a href="http://%1$s%2$s">%1$s%2$s</a>', $signup->domain, $blog_details->path ),
/** 7*/			$signup->user_login,
/** 8*/			$signup->user_email,
/** 9*/			wp_lostpassword_url()
/** 10*/		);

/** 11*/		echo __( 'aaaaaaa' );
/** 12*/		echo __( 'aaaaaaa' );
/** 13 */		echo __( 'aaaaaaa' );
/** 14 */
/** 15 */      printf(
\t\t\t\t/* translators: 1: Site URL, 2: Username, 3: User email address, 4: Lost password URL. */
\t\t\n\t\t__( 'aaaaaaa' ),
/** 18 */\t\t\t\t);`
		const expected = [
			{
				msgid: 'Your site at %1$s is active. You may now log in to your site using your chosen username of &#8220;%2$s&#8221;. Please check your email inbox at %3$s for your password and login instructions. If you do not receive an email, please check your junk or spam folder. If you still do not receive an email within an hour, you can <a href="%4$s">reset your password</a>.',
				raw: [
					'Your site at %1$s is active. You may now log in to your site using your chosen username of &#8220;%2$s&#8221;. Please check your email inbox at %3$s for your password and login instructions. If you do not receive an email, please check your junk or spam folder. If you still do not receive an email within an hour, you can <a href="%4$s">reset your password</a>.',
				],
				reference: '#: filename:5',
				type: 'text_domain',
			},
			{
				msgid: 'aaaaaaa',
				raw: ['aaaaaaa'],
				reference: '#: filename:12',
				type: 'text_domain',
			},
			{
				msgid: 'aaaaaaa',
				raw: ['aaaaaaa'],
				reference: '#: filename:13',
				type: 'text_domain',
			},
			{
				msgid: 'aaaaaaa',
				raw: ['aaaaaaa'],
				reference: '#: filename:14',
				type: 'text_domain',
			},
			{
				msgid: 'aaaaaaa',
				raw: ['aaaaaaa'],
				reference: '#: filename:19',
				type: 'text_domain',
			},
		]

		const result = doTree(content, Php as Parser, filename)

		expect(result).toEqual(expected)
	})
})
