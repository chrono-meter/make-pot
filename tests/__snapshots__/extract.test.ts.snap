// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStrings wp cli should extract from an array of translations 1`] = `
SetOfBlocks {
  "blocks": [
    Block {
      "comments": {
        "reference": [
          "filename.php:3",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Overview",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:4",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Hold down the right button to move the model",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:8",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Rotation",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:9",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Left-click and drag to change the angle",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:13",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Zoom",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:14",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Variable is % and not  {$var}",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
  ],
  "path": "filename.php",
}
`;

exports[`getStrings wp cli should extract translations and comments from code content 1`] = `
SetOfBlocks {
  "blocks": [
    Block {
      "comments": {
        "reference": [
          "filename.php:8",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Hello World",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:13",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "I am being ignored",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:15",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "__",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:16",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "esc_attr__",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:17",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "esc_html__",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:18",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "esc_xml__",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:19",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "_e",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:20",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "esc_attr_e",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:21",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "esc_html_e",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:22",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "esc_xml_e",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": [],
        "flag": "",
        "previous": [],
        "reference": [
          "filename.php:23",
          "filename.php:54",
        ],
        "translator": [],
      },
      "msgctxt": "_x_context",
      "msgid": "_x",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:24",
        ],
        "translator": undefined,
      },
      "msgctxt": "_ex_context",
      "msgid": "_ex",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:25",
        ],
        "translator": undefined,
      },
      "msgctxt": "esc_attr_x_context",
      "msgid": "esc_attr_x",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:26",
        ],
        "translator": undefined,
      },
      "msgctxt": "esc_html_x_context",
      "msgid": "esc_html_x",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:27",
        ],
        "translator": undefined,
      },
      "msgctxt": "esc_xml_x_context",
      "msgid": "esc_xml_x",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": [],
        "flag": "",
        "previous": [],
        "reference": [
          "filename.php:28",
          "filename.php:55",
        ],
        "translator": [],
      },
      "msgctxt": undefined,
      "msgid": "_n_single",
      "msgid_plural": "_n_plural",
      "msgstr": [
        "",
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": [],
        "flag": "",
        "previous": [],
        "reference": [
          "filename.php:29",
          "filename.php:56",
        ],
        "translator": [],
      },
      "msgctxt": "foo-plugin",
      "msgid": "_nx_single",
      "msgid_plural": "_nx_plural",
      "msgstr": [
        "",
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:30",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "_n_noop_single",
      "msgid_plural": "_n_noop_plural",
      "msgstr": [
        "",
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:31",
        ],
        "translator": undefined,
      },
      "msgctxt": "_nx_noop_context",
      "msgid": "_nx_noop_single",
      "msgid_plural": "_nx_noop_plural",
      "msgstr": [
        "",
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:34",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "_",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:37",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "_c",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:38",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "_nc_single",
      "msgid_plural": "_nc_plural",
      "msgstr": [
        "",
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:39",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "__ngettext_single",
      "msgid_plural": "__ngettext_plural",
      "msgstr": [
        "",
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:40",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "__ngettext_noop_single",
      "msgid_plural": "__ngettext_noop_plural",
      "msgstr": [
        "",
        "",
      ],
    },
    Block {
      "comments": {
        "extracted": [],
        "flag": "",
        "previous": [],
        "reference": [
          "filename.php:43",
          "filename.php:58",
        ],
        "translator": [],
      },
      "msgctxt": undefined,
      "msgid": "wrong-domain",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:60",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Hello world",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
  ],
  "path": "filename.php",
}
`;

exports[`getStrings wp cli should extract translations and comments from code content 2`] = `
SetOfBlocks {
  "blocks": [
    Block {
      "comments": {
        "reference": [
          "filename.php:4",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Foo Bar",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:7",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Bar Baz",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:10",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "WordPress",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:14",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Spacey text",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:19",
        ],
        "translator": [
          "Long comment
      spanning multiple
      lines",
        ],
      },
      "msgctxt": undefined,
      "msgid": "Short text",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:22",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "Hello JSX",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:26",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "wp.i18n.__",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:27",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "wp.i18n._n_single",
      "msgid_plural": "wp.i18n._n_plural",
      "msgstr": [
        "",
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:30",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "translate.__",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:45",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "foo-plugin",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:52",
        ],
        "translator": [
          "Translators 1!",
        ],
      },
      "msgctxt": undefined,
      "msgid": "hello world",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:55",
        ],
        "translator": [
          "Translators 2!",
        ],
      },
      "msgctxt": undefined,
      "msgid": "foo",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:58",
        ],
        "translator": [
          "localized date and time format, see https://secure.php.net/date",
        ],
      },
      "msgctxt": undefined,
      "msgid": "F j, Y g:i a",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:61",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "on",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:67",
        ],
        "translator": [
          "/*
       * Translators: If there are characters in your language that are not supported
       * by Lato, translate this to 'off'. Do not translate into your own language.
       */",
        ],
      },
      "msgctxt": undefined,
      "msgid": "off",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:69",
        ],
        "translator": [
          "this should get extracted.",
        ],
      },
      "msgctxt": undefined,
      "msgid": "baba",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:71",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "bubu",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:76",
        ],
        "translator": undefined,
      },
      "msgctxt": undefined,
      "msgid": "yolo",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:79",
        ],
        "translator": [
          "This is a comment",
        ],
      },
      "msgctxt": undefined,
      "msgid": "Plugin name",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
    Block {
      "comments": {
        "reference": [
          "filename.php:82",
        ],
        "translator": [
          "This is another comment!",
        ],
      },
      "msgctxt": undefined,
      "msgid": "https://example.com",
      "msgid_plural": undefined,
      "msgstr": [
        "",
      ],
    },
  ],
  "path": "filename.php",
}
`;
